sled-agent/config-reconciler/src/host_phase_2.rs

```diff
@@ -71,29 +77,270 @@ pub enum BootPartitionError {
         expected: ArtifactHash,
         got: ArtifactHash,
     },
+    #[error("failed to {path} for writing")]
+    OpenForWriting {
+        path: Utf8PathBuf,
+        #[source]
+        err: io::Error,
+    },
+    #[error("could not find desired artifact {desired}")]
+    MissingDesiredArtifact {
+        desired: ArtifactHash,
+        #[source]
+        err: anyhow::Error,
+    },
+    #[error("failed to write to path {path}")]
+    Write {
+        path: Utf8PathBuf,
+        #[source]
+        err: io::Error,
+    },
+    #[error("failed to finalize writes to path {path}")]
+    FinalizeWrite {
+        path: Utf8PathBuf,
+        #[source]
+        err: io::Error,
+    },
 }
 
-#[derive(Debug)]
-pub struct BootPartitionContents {
-    pub boot_disk: Result<M2Slot, BootDiskNotFound>,
-    pub slot_a: Result<BootPartitionDetails, BootPartitionError>,
-    pub slot_b: Result<BootPartitionDetails, BootPartitionError>,
+#[derive(Debug, Default)]
+pub(crate) struct BootPartitionReconciler {
+    // Any time the config reconciler runs, we want to report the contents of
+    // our boot partitions, which requires reading and hashing a full OS image.
+    // This isn't super expensive, but does take a handful of seconds, and it's
+    // really easy for us to cache previous results.
+    //
+    // Cache invalidation is hard; this assumes:
```

Comment:
I wonder if it would be worth it to have a periodic background task that hashes the contents and can be used to compare with the expected value. This would primarily be used for detecting malicious or buggy writes, and certainly isn't a priority now. But it may be useful in the future.